### PR TITLE
Bug#3531

### PR DESCRIPTION
**Pull request by jkerry**
Fix for issue: #3531 registry_key resource is case sensitive in chef but not on windows

* Changed the registry_key provider to hash downcased keys and compare with downcased key names to prevent unnecessary convergence.
* Changed the Win32.Registry class to compare keys and value names with a case insensitive comparison.
* Added functional and unit tests to cover these changes

This was manually tested locally with the recipe described in  #3531

### CODE REVIEW

File: lib/chef/win32/registry.rb [hunk 1]
```diff
@@ -199,11 +199,19 @@ def registry_system_architecture
         ( applied_arch == :x86_64 ) ? 0x0100 : 0x0200
       end
 
+      def SafelyDowncase(val)
```
Comment (jdmundrawala) [inline lib/chef/win32/registry.rb]: we use snake_case

File: lib/chef/win32/registry.rb [hunk 2]
```diff
@@ -199,11 +199,19 @@ def registry_system_architecture
...
+        begin
+          return (val == nil ? nil : val.downcase)
+        rescue NoMethodError
```
Comment (jdmundrawala) [inline lib/chef/win32/registry.rb]: I'm not a fan of expecting such an exception.

**Comment by jkerry** [comment]
Doh! A c# developer's first crack at Ruby. I'll get the method name fixed and look into the exception. It's possible to cast the object to a string first but that seems dirtier.

**Comment by jkerry** [comment]
Corrected both comments.  I'm not sure if there's a more idiomatic ruby way to safely downcase in this situation. If there is please let me know and I'll correct it

**Comment by jdmundrawala** [comment]
:+1: LGTM
cc @chef/client-windows

**Comment by ksubrama** [comment]
:+1: LGTM as well.  Thank you!

**Comment by pburkholder** [comment]
@jkerry :dancer: :balloon: Congrats on first Chef PR.

**Comment by jkerry** [comment]
ht_sam_griner_then_02_jef_150413_16x9_992

**Comment by ranjib** [comment]
:+1:

**Comment by ksubrama** [comment]
Merged at 8c09a658cffd8ff8c654aab1d8ccff69342737c7

**Pull request closed without merging**